feat(hud): Head-Up Display for A32NX and A380X

nonenty opened the pull request:
<!-- ⚠⚠ Do not delete this pull request template! ⚠⚠ -->
<!-- Pull requests that do not follow this template are likely to be ignored. -->

<!-- Add the issues this PR fixes here. If no issues are related to this PR, then this line can be removed. -->
<!-- Add further issues with a full "Fixes #[issue_no]" line to ensure GitHub closes each one when the PR is merged. -->

Fixes #[3415]

## Summary of Changes

<!-- Please provide a summary of changes for this pull request, ensuring all changes are explained. -->

This pull request intends to implement HUD on A32NX and A380X. The code is modified based on PFD.

- [ ] model modifications
  - [x] simple HUD panel
  - [ ] HUD control panel
  - [ ] precise HUD panel with animations
- [ ] Attitude
  - [x] Horizon line
  - [x] Aircraft reference
  - [ ] Roll
  - [x] Pitch/FPA scale
  - [x] Sideslip index
  - [ ] Tail strike pitch limit
- [ ] Airspeed
  - [x] Speed tape
  - [x] Speed trend arrow
  - [ ] Airspeed target
  - [x] Mach number
  - [x] Airspeed indications(V1, V2 etc.)
  - [x] Airspeed protections
- [ ] Altitude
  - [ ] Altitude tape
  - [ ] Altitude target
  - [ ] Vertical deviation
  - [ ] Metric indication
  - [ ] Baro
- [ ] Radio altitude
- [ ] Vertical speed
- [ ] Heading/Track
  - [ ] Heading/Track scale
  - [ ] Current heading/track
  - [ ] Selected heading/track
  - [ ] True indication
- [x] Guidance
  - [x] Flight path vector
  - [x] Total flight path angle
  - [x] Flight path director
  - [x] Delta speed
  - [x] Selected flight path angle
- [ ] Approach guidance
  - [ ] Flare reminder
  - [ ] LOC and G/S deviation
  - [x] Synthetic runway
  - [ ] Information
- [ ] FMA
- [ ] Wind indications
- [ ] Various messages
- [ ] Failure display
- [ ] Display modes
  - [ ] On ground indications
    - [ ] Taxi mode
    - [ ] Takeoff mode
    - [ ] Roll out or rejected takeoff mode
  - [ ] Reduced display modes
    - [ ] Declutter modes
    - [ ] Crosswind mode

## Screenshots (if necessary)

Still WIP.

![Microsoft Flight Simulator 2024_3_22 17_02_59](https://github.com/flybywiresim/aircraft/assets/45462394/6cfafaa4-6b0c-4ebb-87ae-05072cb911b4)

<!-- If your PR includes visual changes, screenshots from before and after your change should always be included. -->
<!-- Please make your best efforts to provide useful before and after screenshots. They should match camera angle, zoom, size, time of day, etc. -->

## References

<!-- You should be making changes based on some kind of a reference (manuals, videos, IRL photos). P3D/xplane/fsx references will only be accepted if we believe that one cannot reasonably obtain a better source. Please post screenshots of the references you used. Ask around in the discord for how to find references for what you are working on. Exceptions will probably be made for IRL A320 pilots and engineers. -->

<!-- If you are making a pull request related to the MCDU, please make sure you are ONLY referencing the Honeywell Pegasus Step 1A (Rev 0), 2009 edition manual. -->
<!-- If you do not have this manual, please ask on our discord for assistance -->

| link | type | date | content |
| ---- | ---- | ---- | ---- |
| [Head-Up Display: see the sky with new eyes](https://www.youtube.com/watch?v=QPCCOeV4KIU) | A320 FFS | 2016.7.29 | release Head-up Combiner Unit(HCU), takeoff, descent, go around, landing, rollout |
| [Live from the Airbus Cockpit - A beautyfull approach into Keflavik Iceland](https://www.youtube.com/watch?v=ARh7bnCfFyk) | A319-115 REG/OY-RCG | 2018.1.30 | descent, approach, landing, rollout, declutter mode |
| [Live from the Airbus Cockpit - Why we fly - Head up Display approach](https://www.youtube.com/watch?v=tqhC2fqBwGk) | Maybe the same above | 2017.12.23 | descent, approach, landing, rollout|
| KAL A380 FLEET FCOM | A380 | 2011.12.03 | HUD system description|
| [350-941 FLEET FCOM](https://github.com/Eyderoe/FlightOperationsManual/tree/main/%E7%A9%BA%E5%AE%A2A350-900)| A350-941 | 2020.2.6 | HUD system description, Chinese version |

## Additional context

<!-- Add any other context about the pull request here. -->

A simple HUD panel is added to the A32NX cockpit for development purposes. Expect further work from modelers.

<!-- You may optionally provide your discord username, so that we may contact you directly about the issue. -->

Discord username (if different from GitHub): nobodyimmortal

## Testing instructions

<!-- Detail how this PR should be tested by QA. Try to list important items that need checking, either directly changed by this PR or that could be affected by it -->

<!-- DO NOT DELETE THIS -->

## How to download the PR for QA

Every new commit to this PR will cause new A32NX and A380X artifacts to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on either **flybywire-aircraft-a320-neo** or **flybywire-aircraft-a380-842** download link at the bottom of the page

nonenty committed:
add hud model

nonenty committed:
add HUD forked from PFD

nonenty committed:
add A32NX_HUD_DECLUTTER_MODE

nonenty committed:
update AircraftReference symbol

nonenty committed:
Update FDYawBar symbol

nonenty committed:
update FPD and FPV

nonenty committed:
pitch scale and horizon line

BlueberryKing commented:
Hi, great to see you are interested in contributing! Just so you're aware, there's currently some discussion within the team about how we plan on sharing code for the PFD between the A32NX and the A380X since we don't want to have to maintain what is mostly the same codebase in two different places. 

As a result, we're not sure if we want to have another instance of largely the same PFD code in the form of the HUD before we've decided on a better structure. I don't want to discourage you from continuing to work on this but wanted to make you aware that you might have to restructure your code if we end up splitting the PFD into reusable components for example.

Unfortunately, I can't say when or even if that's gonna happen, but I want to be fully transparent with you. Other team members are welcome to share their thoughts of course.

nonenty committed:
fix pitch scale render error

nonenty committed:
update alt indicator

nonenty committed:
update speed indicator

nonenty committed:
fix MachNumber display error

nonenty committed:
fix rollGroup rotation center

nonenty committed:
fix FPD roll order display

nonenty commented:
It's just a problem that makes me hesitate recently. I notice that there's pillow distortion when MSFS renders the HUD. Not sure about how to correct it. <feDisplacementMap> may help but too tricky.
![Microsoft Flight Simulator 2024_3_30 20_27_10](https://github.com/flybywiresim/aircraft/assets/45462394/016bd268-63d2-42f2-80d2-08c3e39384d1)
![Microsoft Flight Simulator 2024_3_30 16_47_57](https://github.com/flybywiresim/aircraft/assets/45462394/3dcda6a6-4b13-43e5-940a-ae138df24965)
![image](https://github.com/flybywiresim/aircraft/assets/45462394/9ed65c86-a21d-419d-9510-4767c5214f61)

tracernz commented:
Try render your test pattern on the default B787 HUD (edit it's panel.cfg), and use the pre-defined HUD camera to view it.

nonenty committed:
update HUD size

nonenty committed:
update Attitude Indicator Horizon to match new HUD size

nonenty committed:
fix 'speedAr' topic no instance error

nonenty committed:
Add HUD synthetic runway feature

nonenty committed:
fix rotation center

nonenty committed:
fix FPV visibility condition
update aircraft reference symbol and its visibility condition

nonenty committed:
update HUD pitch_offset

nonenty committed:
fix var names in HUDSimvarPublisher

nonenty committed:
Update FlightPathVector.tsx

move FlightPathDirector to FlightPathVector

implement TotalFlightPathAngle, SelectedFlightPathAngle and DeltaSpeed

nonenty committed:
Merge branch 'flybywiresim:master' into master

nonenty committed:
Merge branch 'master' of https://github.com/nonenty/a32nx into master

Zoverr commented:
news ?

6639835 commented:
Check https://github.com/flybywiresim/aircraft/pull/9428
